File: java/classic/greedy_algorithm/HuffmanCoding.java
package classic.greedy_algorithm;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;

public class HuffmanCoding {
    public String solution(String text) {
        List<Character> aux = Arrays.asList(text.chars().mapToObj(c -> (char) c).toArray(Character[]::new));
        int maxValue = Collections.max(aux);
        int minValue = Collections.min(aux);
        TreeNode[] arrays = new TreeNode[maxValue - minValue + 1];
        Arrays.fill(arrays, null);

        TreeNode current = getRoot(text, maxValue - minValue + 1, minValue, arrays);
        Stack<TreeNode> stack = new Stack<>();
        while (null != current || !stack.isEmpty()) {
            while (null != current) {
                String code = current.code.toString();
                stack.push(current);

                current = current.left;
                if (null != current) {
                    current.code.append(code).append("0");
                }
            }

            current = stack.pop();
            String code = current.code.toString();

            current = current.right;
            if (null != current) {
                current.code.append(code).append("1");
            }
        }

        StringBuffer result = new StringBuffer();
        for (char value : text.toCharArray()) {
            result.append(arrays[value - minValue].code);
        }
        return result.toString();
    }

    private TreeNode getRoot(String text, int size, int minValue, TreeNode[] arrays) {
        int[] bucktes = new int[size];
        Arrays.fill(bucktes, 0);
        for (char value : text.toCharArray()) {
            ++bucktes[value - minValue];
        }

        Queue<TreeNode> queue = new PriorityQueue<>(cmp);
        for (int i = 0; i < bucktes.length; ++i) {
            TreeNode node = new TreeNode(bucktes[i], null, null);
            queue.offer(node);
            arrays[i] = node;
        }

        while (queue.size() > 1) {
            TreeNode left = queue.poll();
            TreeNode right = queue.poll();
            queue.offer(new TreeNode(left.weight + right.weight, left, right));
        }
        return queue.poll();
    }

    private class TreeNode {
        public int weight = 0;
        public TreeNode left = null;
        public TreeNode right = null;
        public StringBuffer code = new StringBuffer();

        public TreeNode(int weight, TreeNode left, TreeNode right) {
            this.weight = weight;
            this.left = left;
            this.right = right;
        }
    }

    private static Comparator<TreeNode> cmp = new Comparator<TreeNode>() {
        @Override
        public int compare(TreeNode node1, TreeNode node2) {
            return node1.weight - node2.weight;
        }
    };
}

File: py/classic/greedy_algorithm/HuffmanCoding.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import heapq
from typing import List

class HuffmanCoding:
    def solution(self, text: str) -> str:
        maxValue, minValue = ord(max(text)), ord(min(text))
        arrays = [None for _ in range(maxValue - minValue + 1)]

        current, stack = self.__getRoot(text, maxValue - minValue + 1, minValue, arrays), []
        while current or stack:
            while current:
                code = current.code
                stack.append(current)

                current = current.left
                if current:
                    current.code = code + '0'

            current = stack.pop()
            code = current.code

            current = current.right
            if current:
                current.code = code + '1'

        result = ''
        for value in text:
            result += arrays[ord(value) - minValue].code
        return result

    def __getRoot(self, text: str, size: int, minValue: int, arrays: List[int]) -> '__TreeNode':
        buckets = [0 for _ in range(size)]
        for value in text:
            buckets[ord(value) - minValue] += 1

        queue = []
        for i, value in enumerate(buckets):
            node = self.__TreeNode(value, None, None)
            heapq.heappush(queue, node)
            arrays[i] = node

        while len(queue) > 1:
            left, right = heapq.heappop(queue), heapq.heappop(queue)
            heapq.heappush(queue, self.__TreeNode(left.weight + right.weight, left, right))
        return queue.pop()

    class __TreeNode:
        def __init__(self, weight: int, left: '__TreeNode', right: '__TreeNode'):
            self.weight, self.left, self.right, self.code = weight, left, right, ''

        def __lt__(self, other: '__TreeNode'):
            return self.weight < other.weight
